// @pega/cosmos-react-core/lib/components/Dialog/Dialog.styles.js
import styled, { css } from 'styled-components';
import { defaultThemeProp } from '../../theme';
import Popover from '../Popover';
export const StyledChildrenWrap = styled.div ``;
export const StyledDialogHeader = styled.div ``;
export const StyledDialogContent = styled.div(({ theme }) => {
    return css `
    position: relative;
    overflow: auto;
    padding: calc(2 * ${theme.base.spacing});

    /* Small amount of block padding accounts for focus ring clipping within overflow auto. */
    ${StyledDialogHeader} + & {
      padding-block-start: calc(0.5 * ${theme.base.spacing});
    }

    &:not(:last-child) {
      padding-block-end: calc(0.5 * ${theme.base.spacing});
    }
  `;
});
StyledDialogContent.defaultProps = defaultThemeProp;
export const StyledDialogFooter = styled.div ``;
export const StyledDialog = styled(Popover)(({ theme }) => {
    return css `
    display: flex;
    flex-direction: column;

    /*
     * For fixed position, which most popovers should be, 100% will be relative to viewport and avoid an annoying 100vw scrollbar nuance.
     * https://stackoverflow.com/questions/29551606/why-does-vw-include-the-scrollbar-as-part-of-the-viewport
     */
    min-width: min(${theme.base['content-width'].sm}, calc(100% - 2rem));
    max-width: min(${theme.base['content-width'].lg}, calc(100% - 2rem));
    max-height: calc(100% - 2rem);

    &[aria-busy='true'] {
      ${StyledChildrenWrap} {
        visibility: hidden;
      }
    }
  `;
});
StyledDialog.defaultProps = defaultThemeProp;
