// @pega/cosmos-react-core/lib/components/MultiStepForm/MultiStepForm.js
import { jsx as _jsx, Fragment as _Fragment, jsxs as _jsxs } from "react/jsx-runtime";
import { forwardRef, useRef, useEffect } from 'react';
import styled, { css } from 'styled-components';
import { useUID, useI18n, usePrevious, useBreakpoint, useTheme } from '../../hooks';
import { getFocusables } from '../../utils';
import { calculateFontSize } from '../../styles';
import Flex from '../Flex';
import Text from '../Text';
import HTML from '../HTML';
import VisuallyHiddenText from '../VisuallyHiddenText';
import Grid from '../Grid';
import { defaultThemeProp } from '../../theme';
import VerticalFormProgress from './VerticalFormProgress';
import HorizontalFormProgress from './HorizontalFormProgress';
export const StyledForm = styled.form(({ theme }) => {
    const { base: { spacing } } = theme;
    return css `
    height: 100%;
    width: 100%;
    flex-grow: 1;
    overflow-y: auto;
    padding: ${spacing};
  `;
});
export const StyledFormContent = styled.div ``;
export const StyledRequiredFieldLegend = styled(Text)(({ theme }) => {
    const { base: { 'font-size': fontSize, 'font-scale': fontScale, palette: { urgent } } } = theme;
    const { xxs: infoFontSize } = calculateFontSize(fontSize, fontScale);
    return css `
    font-size: ${infoFontSize};
    font-style: italic;

    &::before {
      display: 'inline';
      content: '\\00a0*';
      vertical-align: top;
      color: ${urgent};
    }
  `;
});
StyledRequiredFieldLegend.defaultProps = defaultThemeProp;
export const MultiStepForm = forwardRef(({ currentStepId, steps, heading, stepIndicator: stepIndicatorProp = steps.length < 2 ? 'none' : 'horizontal', ...restProps }, ref) => {
    const multiStepContentRef = useRef(null);
    const previousId = usePrevious(currentStepId);
    const currentStep = steps.find(step => step.id === currentStepId);
    const headingId = useUID();
    const descriptionId = useUID();
    const t = useI18n();
    const { base: { 'content-width': contentWidth } } = useTheme();
    useEffect(() => {
        if (previousId === undefined)
            return;
        if (previousId !== currentStepId) {
            const focusableElements = getFocusables(multiStepContentRef);
            if (focusableElements.length)
                focusableElements[0].focus();
        }
    }, [currentStepId, previousId]);
    const isSmallOrAbove = useBreakpoint('sm');
    const stepIndicator = stepIndicatorProp === 'vertical' && !isSmallOrAbove ? 'horizontal' : stepIndicatorProp;
    let areaDef = '';
    switch (stepIndicator) {
        case 'horizontal':
            if (heading)
                areaDef += '"heading" ';
            areaDef += '"stepIndicator" "content"';
            if (currentStep.actions)
                areaDef += ' "actions"';
            break;
        case 'vertical':
            if (heading)
                areaDef += '"heading stepIndicator" ';
            areaDef += '"content stepIndicator"';
            if (currentStep.actions)
                areaDef += ' "actions actions"';
            break;
        default:
            if (heading)
                areaDef += '"heading" ';
            areaDef += '"content"';
            if (currentStep.actions)
                areaDef += ' "actions"';
    }
    return (_jsxs(Grid, { ref: ref, as: StyledForm, ...restProps, container: {
            cols: stepIndicator === 'vertical' ? `2fr minmax(${contentWidth.xs}, 1fr)` : 'minmax(0, 1fr)',
            areas: areaDef,
            gap: 3
        }, "aria-labelledby": heading ? headingId : undefined, "aria-describedby": currentStep.description ? descriptionId : undefined, children: [(heading || steps.length > 1) && (_jsxs(_Fragment, { children: [heading && (_jsx(Grid, { item: { area: 'heading' }, children: _jsx(Text, { id: headingId, variant: 'h3', children: heading }) })), steps.length > 1 && (_jsxs(_Fragment, { children: [stepIndicator === 'horizontal' && (_jsx(Grid, { item: { area: 'stepIndicator' }, as: HorizontalFormProgress, steps: steps, currentStepId: currentStepId })), stepIndicator === 'vertical' && (_jsx(Grid, { item: { area: 'stepIndicator' }, as: VerticalFormProgress, steps: steps, currentStepId: currentStepId })), _jsx(VisuallyHiddenText, { "aria-live": 'polite', children: previousId !== undefined && previousId !== currentStepId
                                    ? t('step_changed_to_name', [currentStep.name])
                                    : '' })] }))] })), _jsxs(Grid, { item: { area: 'content' }, container: { gap: 3 }, children: [currentStep.banners, currentStep.description && (_jsx(HTML, { id: descriptionId, as: 'p', content: currentStep.description })), _jsx(StyledFormContent, { ref: multiStepContentRef, children: currentStep.content })] }), currentStep.actions && (_jsx(Grid, { item: { area: 'actions' }, children: _jsx(Flex, { container: { justify: 'between', alignItems: 'center' }, children: currentStep.actions }) }))] }));
});
export default MultiStepForm;
